### ISSUE TITLE
Some SIMD tests fail on AVX2 machine under GCStress=C

### ISSUE THREAD

**Issue by RussKeldorph** [dotnet/coreclr#5759]
**Environment**
Ubuntu x64 Checked (machine **MUST** support AVX2)
COMPlus_GCStress=C

**Sample Output (other tests fail in different ways)**
```
/mnt/j/workspace/dotnet_coreclr/master/x64_checked_ubuntu_gcstress0xc_tst/bin/tests/Windows_NT.x64.Checked/Tests/coreoverlay/corerun DivSignedUnsignedTest_r.exe

Unhandled Exception: System.DivideByZeroException: Attempted to divide by zero.
    at System.Numerics.Vector`1.ScalarDivide(T left, T right)
    at System.Numerics.Vector`1.op_Division(Vector`1 left, Vector`1 right)
    at VectorMathTests.Program.Main(String[] args)
./DivSignedUnsignedTest_r.sh: line 121: 31700 Aborted                 (core dumped) $_DebuggerFullPath "$CORE_ROOT/corerun" DivSignedUnsignedTest_r.exe $CLRTestExecutionArguments
Expected: 100
Actual: 134
END EXECUTION - FAILED
```

**Sample CI Job with failures**
https://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/x64_checked_ubuntu_gcstress0xc_tst/14/testReport/

**List of failing tests in above CI Job**
```
JIT.SIMD.BoxUnbox_r.BoxUnbox_r
JIT.SIMD.CreateGeneric_r.CreateGeneric_r
JIT.SIMD.CreateGeneric_ro.CreateGeneric_ro
JIT.SIMD.CtorFromArray_r.CtorFromArray_r
JIT.SIMD.CtorFromArray_ro.CtorFromArray_ro
JIT.SIMD.DivSignedUnsignedTest_r.DivSignedUnsignedTest_r
JIT.SIMD.Haar-likeFeaturesGeneric_r.Haar-likeFeaturesGeneric_r
JIT.SIMD.Haar-likeFeaturesGeneric_ro.Haar-likeFeaturesGeneric_ro
JIT.SIMD.StoreElement_r.StoreElement_r
JIT.SIMD.VectorAbs_ro.VectorAbs_ro
JIT.SIMD.VectorAbs_r.VectorAbs_r
JIT.SIMD.VectorAdd_r.VectorAdd_r
JIT.SIMD.VectorArgs_ro.VectorArgs_ro
JIT.SIMD.VectorArgs_r.VectorArgs_r
JIT.SIMD.VectorArrayInit_ro.VectorArrayInit_ro
JIT.SIMD.VectorArrayInit_r.VectorArrayInit_r
JIT.SIMD.VectorArray_ro.VectorArray_ro
JIT.SIMD.VectorArray_r.VectorArray_r
JIT.SIMD.VectorDiv_r.VectorDiv_r
JIT.SIMD.VectorDot_r.VectorDot_r
JIT.SIMD.VectorExp_ro.VectorExp_ro
JIT.SIMD.VectorExp_r.VectorExp_r
JIT.SIMD.VectorGet_r.VectorGet_r
JIT.SIMD.VectorHWAccel2_r.VectorHWAccel2_r
JIT.SIMD.VectorHWAccel_r.VectorHWAccel_r
JIT.SIMD.VectorIntEquals_r.VectorIntEquals_r
JIT.SIMD.VectorMatrix_ro.VectorMatrix_ro
JIT.SIMD.VectorMatrix_r.VectorMatrix_r
JIT.SIMD.VectorMax_r.VectorMax_r
JIT.SIMD.VectorMin_r.VectorMin_r
JIT.SIMD.VectorMul_r.VectorMul_r
JIT.SIMD.VectorRelOp_ro.VectorRelOp_ro
JIT.SIMD.VectorRelOp_r.VectorRelOp_r
JIT.SIMD.VectorSub_r.VectorSub_r
JIT.SIMD.VectorReturn_r.VectorReturn_r
```

**Comment by RussKeldorph** [comment]
@CarolEidt @sivarv @dotnet/jit-contrib FYI

**Comment by RussKeldorph** [comment]
We currently think the problem is in GCStress context save/restore, i.e. a problem with the GCStress implementation and not end-user facing.  Still investigating, however.

**Comment by RussKeldorph** [comment]
No update yet.  Still consider this a risk until it is understood.  I would not hold anything for it right now, though.

**Comment by cmckinsey** [comment]
Could LLVM be using AVX code internally in the code generated for the JIT/runtime that could be trashing AVX register state of the running thread?

**Comment by swaroop-sridhar** [comment]
CC @kouvel

**Comment by swaroop-sridhar** [comment]
This problem clearly happens when GCStress runs for a trap inserted at a non-call instruction (because YMM regs are volaltile across calls).

I changed gccover.cpp to insert traps only at call-sites and fully-interruptible instructions that are CALL-instructions. The tests pass in this mode.

**Comment by swaroop-sridhar** [comment]
The problem here is that:
1)  We enter a GCStress through common_signal_handler() 
2) The trap mechanism has cleared all XMM/YMM registers
3) RtlRegister context captures the register context, but not YMMs
4) After GCStress, RedirectedHandledJITCase() simply returns via RtlRestoreContext() thus not restoring the YMM registers.

**Comment by swaroop-sridhar** [comment]
This problem only happens when we enter GC via a trap (i.e. for GCStress).
So, we need to fix the GCStress handlers to capture the full context (on machines that are AVX-enabled, have non-zero AVX context, and are not at a call-site).

**Comment by janvorli** [comment]
@swaroop-sridhar - you say that the trap mechanism has cleared all XMM/YMM registers. If that's the case, then the only way to capture / restore them would be to find a way to extract them from / patch them in the unix context.
That needs to happen in CONTEXTFromNativeContext / CONTEXTToNativeContext. And of course we need to capture / restore them in the RtlCaptureContext / RtlRestoreContext too.

**Comment by kouvel** [comment]
Fixed by #6394
